### PR TITLE
Fix parsing of char columns in foreign tables result set

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Fixes regression introduced in https://github.com/crate/crate/commit/70ddc1f6eab604ab4220b15c81dfd19a33c621dc

### CODE REVIEW

**Review by mfussenegger — APPROVED** [review]

**Comment by BaurzhanSakhariev** [comment]
https://www.postgresql.org/docs/15/release-15.html

>Create a new [pg_type.typcategory](https://www.postgresql.org/docs/15/catalog-pg-type.html) value for "char" (Tom Lane)

File: server/src/main/java/io/crate/types/ResultSetParser.java [hunk 1]
```diff
@@ -80,11 +80,12 @@ public static Object getObject(ResultSet resultSet, int i, String typeName) thro
                 return PgOidVectorType.listFromOidVectorString(textval);
             }
             case "char":
-                String strValue = resultSet.getString(columnIndex);
-                if (strValue == null) {
+                // It's not a single byte type (CharType) as char is not surrounded by quotes.
+                var str = resultSet.getString(columnIndex);
+                if (str == null) {
                     return null;
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/types/ResultSetParser.java:86]: flaky tests - `testAnyOnArrayLiteral` with `UseJdbc = 1 `fails consistently.

We have to differentiate here between `char` and `"char"`. Problem is that driver uses OID to resolve type name and OID is correctly assigned to 18 in CharType. In postgres pg_catalog.pg_type doesn't contain special entry for "char".

**Comment by BaurzhanSakhariev** [comment]
This fix is wrong. We shouldn't treat "char" neither as string (this PR) nor as numeric only Byte (master). 
As stated in https://www.postgresql.org/docs/current/datatype-character.html, "char" can store single ASCII character.
> The type "char" (note the quotes) is different from char(1) in that it only uses one byte of storage, and therefore can store only a single ASCII character. 

I debugged `ResultSetMetaData.getColumnTypeName`, it returns "char" for Postgres `prokind`, resolved internally by 
OID which is equal to 18.

Another way to check `prokind` type is
```
select column_name, data_type from information_schema.columns
where table_name = 'pg_proc' and column_name = 'prokind';
```

output: "char" (with quotes).


The only confusing  part in PG is that they didn't bother surrounfing prokind type with quotes in pg_proc docs
https://www.postgresql.org/docs/current/catalog-pg-proc.html

**Comment by BaurzhanSakhariev** [comment]
I think correct fix is reading 1 byte and accept non-numeric symbols. 

Probably, it also indicates that our `CharType` shouldn't be backed by numeric only Byte (separate issue)

**Comment by BaurzhanSakhariev** [comment]
> We shouldn't treat "char" neither as string (this PR) nor as numeric only Byte (master).

Actually it's okay but I pushed a commit to do it in a different way to ensure that we use ASCII and assert on length. 

@mfussenegger  could you please take another look?

File: server/src/main/java/io/crate/types/ResultSetParser.java [hunk 2]
```diff
@@ -80,11 +80,16 @@ public static Object getObject(ResultSet resultSet, int i, String typeName) thro
...
+                // It's char with OID 18, ie single byte "char" which can store single ASCII character.
+                // See also https://www.postgresql.org/docs/15/datatype-character.html#DATATYPE-CHARACTER-SPECIAL-TABLE
+                byte[] charBytes = resultSet.getBytes(columnIndex);
+                if (charBytes == null) {
                     return null;
                 }
-                return Byte.valueOf(strValue);
+                String str = new String(charBytes, StandardCharsets.US_ASCII);
+                assert str.length() == 1 : "Only single ASCII character is allowed";
+                return str;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/ResultSetParser.java:92]: Can we just return `resultSet.getString(columnIndex)` here?
I don't think we need to do additional validation, that's up to the foreign server to ensure that a char type has a value matching its char type semantics

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/types/ResultSetParser.java]: updated

**Review by mfussenegger — APPROVED** [review]

**Comment by BaurzhanSakhariev** [comment]
Test failure was related to missing NULL check in addition to `if (typeName.equals("char")`, fixed

**Pull request merged**